pkg/policy/rule.go

```diff
@@ -58,19 +58,44 @@ func (r *rule) String() string {
 	return fmt.Sprintf("%v", r.EndpointSelector)
 }
 
-func mergePortProto(ctx *SearchContext, endpoints []api.EndpointSelector, existingFilter, filterToMerge *L4Filter) error {
+func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter, selectorCache *SelectorCache) error {
 	// Handle cases where filter we are merging new rule with, new rule itself
 	// allows all traffic on L3, or both rules allow all traffic on L3.
 	//
 	// Case 1: either filter selects all endpoints, which means that this filter
 	// can now simply select all endpoints.
-	if existingFilter.AllowsAllAtL3() || filterToMerge.AllowsAllAtL3() {
-		existingFilter.Endpoints = api.EndpointSelectorSlice{api.WildcardEndpointSelector}
+	if existingFilter.AllowsAllAtL3() {
+		// Release references held by filterToMerge.CachedSelectors
+		selectorCache.RemoveSelectors(filterToMerge, filterToMerge.CachedSelectors)
```

Comment:
I think that this "remove and set to nil" functionality should be put in some utility function since it's crucial we have the two collocated together.